extern/test_bmi_py/bmi_grid.py

```diff
@@ -0,0 +1,200 @@
+        Structure for holding required BMI meta data for any grid intended to be used via BMI
+    """
+    def __init__(self, id: int, rank: int , type: GridType):
+        """_summary_
+
+        Args:
+            id (int): User defined identifier for this grid
+            rank (int): The number of dimensions of the grid
+            type (GridType): The type of BMI grid this meta data represents
+        """
+        self._id: int = id
+        self._rank: int = rank
+        self._size: int = 0
+        self._type: GridType = type #FIXME validate type/rank?
+        self._shape: 'NDArray[np.int32]' = None #array of size rank
+        self._spacing: 'NDArray[np.float64]' = None #array of size rank
+        self._origin: 'NDArray[np.float64]' = None #array of size rank
+        if( rank == 0 ):
+            # We have to use a 1 dim representation for a scalar cause numpy initialization is weird
+            # np.zeros( [1] ) gives you an array([0.])
+            # np.zeros( [0] ) gives you an emptty array([])
+            # np.zeros( () ) gives a scalar wrapped in an array array(0.)
+            # This latter is really what we want, but then it is hard to communicate the actual size
+            # (as a numerical value...)
+            self._shape = np.zeros( (), np.int32 ) #note, int32 is important here -- assumed by ngen
+            #self._shape[...] = 1
+        else:
+            self._shape = np.zeros( rank, np.int32) #set the shape rank, with 0 allocated values
+        #Make the array "immutable", can only modify via setting
+        self._shape.flags.writeable = False
+
+    # TODO consider restricting resetting of grid values after they have been initialized
+
+    @property
+    def id(self) -> int:
+        """The unique grid identifer.
+
+        Returns:
+            int: grid identifier
+        """
+        return self._id
+
+    @property
+    def rank(self) -> int:
+        """The dimensionality of the grid.
+
+        Returns:
+            int: Number of dimensions of the grid.
+        """
+        return self._rank
+
+    @property
+    def size(self) -> int:
+        """The total number of elements in the grid
+
+        Returns:
+            int: number of grid elements
+        """
+        if not self.shape or self.shape.ndim == 0: #it is None or () or np.array( () )
+            return 0
+        else:
+            #multiply the shape of each dimension together
+            return reduce( lambda x, y: x*y, self._shape)
+
+    @property
+    def type(self) -> GridType:
+        """The type of BMI grid.
+
+        Returns:
+            GridType: bmi grid type
+        """
+        return self._type
+    
+    @property
+    def shape(self) -> 'NDArray[np.int32]':
+        """The shape of the grid (the size of each dimension)
+
+        Returns:
+            Tuple[int]: size of each dimension
+        """
+        return self._shape
+    
+    @shape.setter
+    def shape(self, shape: 'NDArray[np.int32]') -> None:
```

Comment:
Is this type hint for the input wrong?

Reply:
I'm pretty sure this is resolved in the latest changes, but let me know if I'm missing something.